### DISCUSSION TITLE
Unable to trigger OnPickerSelectedIndexChanged event in ViewModel

### DISCUSSION THREAD

**Discussion by cluber22** [dotnet/maui#18294] [Q&A] [answered]
I have a picker and I want to listen for the OnChange event. I am using the MVVM Community Toolkit to bind controls.

```
<Picker x:Name="pickerPatrolType"
    Title="{Binding PicPatrolTypeTitle}"
    ItemsSource="{Binding PicPatrolTypes, Mode=TwoWay}"
    ItemDisplayBinding="{Binding GUIValue}"
    SelectedItem="{Binding PicPatrolTypeSelectedItem}"
    SelectedIndex="{Binding PicPatrolTypeIndex, Mode=TwoWay}"
/>

[ObservableProperty]
string picPatrolTypeTitle;

[ObservableProperty]
List<ChoiceList> picPatrolTypes;

[ObservableProperty]
object picPatrolTypeSelectedItem;

[ObservableProperty]
int picPatrolTypeIndex;

public void OnPicPatrolTypeIndexChanged(object sender, EventArgs e)
{
    var picker = (Picker)sender;
    int selectedIndex = picker.SelectedIndex;
    Console.WriteLine($"OnPatrolTypeIndexChanged: {selectedIndex}");
}
```

For whatever reason I cannot seem to get the listener to call. What am I doing wrong? Any suggestions are welcomed!

References I've found which haven't helped:
https://learn.microsoft.com/en-us/dotnet/maui/user-interface/controls/picker
https://learn.microsoft.com/en-us/dotnet/communitytoolkit/mvvm/
https://stackoverflow.com/questions/72562345/net-maui-mvvm-picker-binding-selectedindexchanged-event

**Comment by CodeRedCatto** [comment]
You have to link the callback to the picker. You’re missing 
`SelectedIndexChanged=“OnPicPatrolTypeIndexChanged”`
in your picker Xaml.

**Reply by cluber22** [reply]
Thanks for this, I however get the following error:

Severity	Code	Description	Project	File	Line	Suppression State
Error	XFC0002	EventHandler "OnPicPatrolTypeIndexChanged" with correct signature not found in type "Namespace.RecordPage"

This is erroring on the line I just added in the XML.  SelectedIndexChanged="OnPicPatrolTypeIndexChanged"


```
<Picker x:Name="pickerPatrolType"
    Title="{Binding PicPatrolTypeTitle}"
    ItemsSource="{Binding PicPatrolTypes, Mode=TwoWay}"
    ItemDisplayBinding="{Binding GUIValue}"
    SelectedItem="{Binding PicPatrolTypeSelectedItem}"
    SelectedIndex="{Binding PicPatrolTypeIndex, Mode=TwoWay}"
    SelectedIndexChanged="OnPicPatrolTypeIndexChanged"
/>

```

**Reply by CodeRedCatto** [reply]
Ah, it probably doesn’t like the return type or the args.

Try removing the callback from the xaml, then re-add it allowing it to autocomplete a new callback and generate it in the code behind. The auto generated one should have the correct signature and you can copy that structure to your function.

From a glance, it looks correct to me. Is your function in RecordPage.xaml.cs?

**Reply by cluber22** [reply]
This is probably where I'm going wrong, I have a view model.

RecordPage.xaml.cs
```
public partial class RecordPage : ContentPage
{
    public RecordPage(RecordViewModel vm)
    {
        InitializeComponent();
        BindingContext = vm;
    }
}
```

RecordViewModel.cs
```
    public partial class RecordViewModel : ObservableObject
    {
// This is where i'm trying to put the OnPicPatrolTypeIndexChanged() method
    }
```

How do I make it work in my ViewModel class and not the RecordPage class?

Thanks for your help on this.

**Reply by FM1973** [reply, marked as answer]
Try using the EventToCommandBehavior from the .net MAUI tookit,
[https://learn.microsoft.com/en-us/dotnet/communitytoolkit/maui/behaviors/event-to-command-behavior](https://learn.microsoft.com/en-us/dotnet/communitytoolkit/maui/behaviors/event-to-command-behavior)

**Reply by cluber22** [reply]
@FM1973 This seems to be my solution - thank you.

```
<Picker
   x:Name="pickerPatrolType"
   Title="{Binding PicPatrolTypeTitle}"
   ItemsSource="{Binding PicPatrolTypes, Mode=TwoWay}"
   ItemDisplayBinding="{Binding GUIValue}"
   SelectedItem="{Binding PicPatrolTypeSelectedItem}"
   SelectedIndex="{Binding PicPatrolTypeIndex, Mode=TwoWay}"
    >
    <Picker.Behaviors>
        <toolkit:EventToCommandBehavior
            EventName="SelectedIndexChanged"
            Command="{Binding CheckPatrolTypeSelectedCommand}" />
    </Picker.Behaviors>
</Picker>

[RelayCommand]
void CheckPatrolTypeSelected()
{
    if (PicPatrolTypeIndex != -1)
    {
        // do stuff
    }
    else
    {
        // do other stuff
    }
}
```

@CodeRedCatto Thanks for your help also
